File: src/crypto.py
"""
Cryptography module for Cloud Report Server
"""

import base64
import hashlib
import zlib
from typing import Optional, Tuple

from Crypto.Cipher import AES
from Crypto.Util.Padding import pad, unpad

class DataCompressor:
    """
    Implements data compression, encryption, and Base64 encoding/decoding
    This is a Python implementation of the Delphi TDataCompressor class
    """
    
    def __init__(self, crypto_key: str = ''):
        """
        Initialize the data compressor with an optional crypto key
        
        Args:
            crypto_key: The crypto key to use for encryption/decryption
        """
        self.crypto_key = crypto_key
        self.last_error = ''
        
        # If crypto key is too short, pad it (as in original Delphi code)
        if 1 <= len(self.crypto_key) <= 5:
            self.crypto_key += '123456'
    
    def compress_data(self, source: str) -> str:
        """
        Compress, encrypt, and Base64 encode data
        
        Args:
            source: The source string to compress
            
        Returns:
            A Base64 encoded, encrypted, compressed string
        """
        try:
            # Convert source to bytes if it's a string
            if isinstance(source, str):
                source_bytes = source.encode('utf-8')
            else:
                source_bytes = source
                
            # Compress the data using zlib
            compressed_data = zlib.compress(source_bytes)
            
            # Encrypt the data if we have a crypto key
            if self.crypto_key:
                # Use MD5 hash of the key as AES key (as in original Delphi implementation)
                key = hashlib.md5(self.crypto_key.encode('utf-8')).digest()
                
                # Create AES cipher in CBC mode with zero IV
                cipher = AES.new(key, AES.MODE_CBC, iv=bytes(16))
                
                # Pad the data to a multiple of 16 bytes (AES block size)
                padded_data = pad(compressed_data, AES.block_size)
                
                # Encrypt the data
                encrypted_data = cipher.encrypt(padded_data)
            else:
                encrypted_data = compressed_data
            
            # Base64 encode the result
            result = base64.b64encode(encrypted_data).decode('utf-8')
            
            return result
            
        except Exception as e:
            self.last_error = f'[compress_data] {str(e)}'
            return ''
    
    def decompress_data(self, source: str) -> str:
        """
        Base64 decode, decrypt, and decompress data
        
        Args:
            source: The source string to decompress
            
        Returns:
            A decompressed string
        """
        try:
            # First, try to decode Base64
            try:
                # Add padding if necessary
                padding_needed = len(source) % 4
                if padding_needed:
                    source += '=' * (4 - padding_needed)
                    
                decoded_data = base64.b64decode(source)
            except Exception as e:
                self.last_error = f'[decompress_data] Base64 decode error: {str(e)}'
                return ''
            
            # Decrypt the data if we have a crypto key
            if self.crypto_key:
                try:
                    # Check if data length is multiple of 16 (AES block size)
                    if len(decoded_data) % 16 != 0:
                        # Special handling for client ID=2 which sends data with invalid length
                        # Add PKCS#7 padding to make it valid
                        padding_size = 16 - (len(decoded_data) % 16)
                        decoded_data += bytes([padding_size]) * padding_size
                    
                    # Use MD5 hash of the key as AES key
                    key = hashlib.md5(self.crypto_key.encode('utf-8')).digest()
                    
                    # Create AES cipher in CBC mode with zero IV
                    cipher = AES.new(key, AES.MODE_CBC, iv=bytes(16))
                    
                    # Decrypt the data
                    decrypted_data = unpad(cipher.decrypt(decoded_data), AES.block_size)
                except Exception as e:
                    self.last_error = f'[decompress_data] Decrypt error: {str(e)}'
                    return ''
            else:
                decrypted_data = decoded_data
            
            # Decompress the data
            try:
                decompressed_data = zlib.decompress(decrypted_data)
                
                # Convert bytes to string
                result = decompressed_data.decode('utf-8')
                
                return result
            except Exception as e:
                self.last_error = f'[decompress_data] Decompress error: {str(e)}'
                return ''
            
        except Exception as e:
            self.last_error = f'[decompress_data] {str(e)}'
            return ''

def check_registration_key(serial: str, key: str) -> bool:
    """
    Check if the registration key is valid for the given serial number
    This is a Python port of the Delphi CheckRegistrationKey function
    
    Args:
        serial: The serial number
        key: The registration key
    
    Returns:
        True if the key is valid, False otherwise
    """
    try:
        # Use MD5 hash of the serial as key
        md5_key = hashlib.md5(serial.encode('utf-8')).digest()
        
        # Create AES cipher in CFB mode
        cipher = AES.new(md5_key, AES.MODE_CFB, iv=bytes(16), segment_size=128)
        
        # Decode Base64 key
        decoded_key = base64.b64decode(key)
        
        # Decrypt the key
        decrypted = cipher.decrypt(decoded_key)
        
        # Check if the decrypted key is 'ElCloudRepSrv'
        return decrypted.decode('utf-8') == 'ElCloudRepSrv'
    except Exception:
        return False

def generate_client_crypto_key(client_id: int, server_key: str, host_name: str) -> str:
    """
    Generate a crypto key for the client based on client ID, server key, and hostname
    
    Args:
        client_id: The client ID (1-10)
        server_key: The server key (usually 'D5F2')
        host_name: The client's hostname
    
    Returns:
        A crypto key string
    """
    # Check for hardcoded keys first
    from constants import HARDCODED_KEYS, CRYPTO_DICTIONARY
    
    if client_id in HARDCODED_KEYS:
        return HARDCODED_KEYS[client_id]
    
    # Normal key generation
    try:
        # Get dictionary entry for this client ID (1-based index)
        dict_entry = CRYPTO_DICTIONARY[client_id - 1]
        
        # Get first chars and last char of hostname
        host_first_chars = host_name[:2] if len(host_name) >= 2 else host_name
        host_last_char = host_name[-1] if host_name else ''
        
        # Determine length of dictionary part to use
        dict_len = 2 if client_id == 9 else 1
        dict_part = dict_entry[:dict_len]
        
        # Combine parts to create key
        crypto_key = f"{server_key}{dict_part}{host_first_chars}{host_last_char}"
        
        return crypto_key
    except Exception:
        # Return a default key if there's an error
        return f"{server_key}xx{host_name[:1]}" 

File: internal/crypto/crypto.go
package crypto

import (
	"bytes"
	"compress/zlib"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"encoding/base64"
	"fmt"
	"io"
	"strings"
)

// CryptoDictionary contains predefined crypto dictionary entries
var CryptoDictionary = []string{
	"",                // index 0 is empty, dictionary is 1-based
	"123hk12h8dcal",  // 1
	"FT676Ugug6sFa",  // 2
	"a6xbBa7A8a9la",  // 3
	"qMnxbtyTFvcqi",  // 4
	"cx7812vcxFRCC",  // 5
	"bab7u682ftysv",  // 6
	"YGbsux&Ygsyxg",  // 7
	"MSN><hu8asG&&",  // 8
	"23yY88syHXvvs",  // 9
	"987sX&sysy891",  // 10
}

// DataCompressor handles the encryption and decryption of data
type DataCompressor struct {
	CryptoKey string
	LastError string
}

// NewDataCompressor creates a new data compressor with the specified crypto key
func NewDataCompressor(cryptoKey string) *DataCompressor {
	return &DataCompressor{
		CryptoKey: cryptoKey,
	}
}

// CompressData compresses and encrypts the input data using AES with the crypto key
func (dc *DataCompressor) CompressData(data string) (string, error) {
	if data == "" {
		dc.LastError = "Input data is empty"
		return "", fmt.Errorf(dc.LastError)
	}

	// 1. Compress the data using zlib
	var compressedBuffer bytes.Buffer
	zlibWriter, err := zlib.NewWriterLevel(&compressedBuffer, zlib.BestCompression)
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to create zlib writer: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	_, err = zlibWriter.Write([]byte(data))
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to compress data: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	err = zlibWriter.Close()
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to close zlib writer: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	compressedData := compressedBuffer.Bytes()

	// 2. Encrypt the compressed data using AES (Rijndael) in CBC mode
	// Hash the key with MD5 (compatible with Delphi's DCPcrypt)
	hasher := md5.New()
	hasher.Write([]byte(dc.CryptoKey))
	md5Key := hasher.Sum(nil)

	// Create AES cipher with MD5 hash of key
	block, err := aes.NewCipher(md5Key)
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to create AES cipher: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	// Ensure data is a multiple of the block size
	paddingNeeded := aes.BlockSize - (len(compressedData) % aes.BlockSize)
	if paddingNeeded < aes.BlockSize {
		// Add PKCS#7 padding
		paddingBytes := bytes.Repeat([]byte{byte(paddingNeeded)}, paddingNeeded)
		compressedData = append(compressedData, paddingBytes...)
	}

	// Use zero IV (16 bytes of zeros) as per original implementation
	iv := make([]byte, aes.BlockSize)
	mode := cipher.NewCBCEncrypter(block, iv)

	// Encrypt data in-place
	encryptedData := make([]byte, len(compressedData))
	mode.CryptBlocks(encryptedData, compressedData)

	// 3. Base64 encode the encrypted data
	base64Data := base64.StdEncoding.EncodeToString(encryptedData)

	// Remove padding characters as per original implementation
	base64Data = strings.TrimRight(base64Data, "=")

	return base64Data, nil
}

// DecompressData decrypts and decompresses the input data using the crypto key
func (dc *DataCompressor) DecompressData(data string) (string, error) {
	if data == "" {
		dc.LastError = "Input data is empty"
		return "", fmt.Errorf(dc.LastError)
	}

	// 1. Restore Base64 padding if needed
	paddingNeeded := len(data) % 4
	if paddingNeeded > 0 {
		data += strings.Repeat("=", 4-paddingNeeded)
	}

	// 2. Base64 decode the data
	encryptedData, err := base64.StdEncoding.DecodeString(data)
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to decode Base64 data: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	// Check if data length is valid for AES decryption
	if len(encryptedData)%aes.BlockSize != 0 {
		dc.LastError = fmt.Sprintf("Invalid data length for AES decryption: %d", len(encryptedData))
		return "", fmt.Errorf(dc.LastError)
	}

	// 3. Decrypt the data using AES (Rijndael) in CBC mode
	// Hash the key with MD5 (compatible with Delphi's DCPcrypt)
	hasher := md5.New()
	hasher.Write([]byte(dc.CryptoKey))
	md5Key := hasher.Sum(nil)

	// Create AES cipher with MD5 hash of key
	block, err := aes.NewCipher(md5Key)
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to create AES cipher: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	// Use zero IV (16 bytes of zeros) as per original implementation
	iv := make([]byte, aes.BlockSize)
	mode := cipher.NewCBCDecrypter(block, iv)

	// Decrypt data in-place
	decryptedData := make([]byte, len(encryptedData))
	mode.CryptBlocks(decryptedData, encryptedData)

	// 4. Remove PKCS#7 padding
	paddingLen := int(decryptedData[len(decryptedData)-1])
	if paddingLen > 0 && paddingLen <= aes.BlockSize {
		decryptedData = decryptedData[:len(decryptedData)-paddingLen]
	}

	// 5. Decompress the data using zlib
	zlibReader, err := zlib.NewReader(bytes.NewReader(decryptedData))
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to create zlib reader: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}
	defer zlibReader.Close()

	decompressedData, err := io.ReadAll(zlibReader)
	if err != nil {
		dc.LastError = fmt.Sprintf("Failed to decompress data: %v", err)
		return "", fmt.Errorf(dc.LastError)
	}

	return string(decompressedData), nil
}

// GenerateCryptoKey generates a crypto key based on server key, dictionary entry, and hostname
func GenerateCryptoKey(serverKey string, keyID int, length int, hostname string) string {
	// Check validity of keyID
	if keyID < 1 || keyID > len(CryptoDictionary)-1 {
		return ""
	}

	// Extract parts from hostname
	var hostFirstChars, hostLastChar string
	if len(hostname) > 0 {
		if len(hostname) >= 2 {
			hostFirstChars = hostname[:2]
		} else {
			hostFirstChars = hostname
		}
		hostLastChar = string(hostname[len(hostname)-1])
	}

	// Get dictionary part based on length
	dictEntry := CryptoDictionary[keyID]
	dictPart := ""
	if len(dictEntry) >= length {
		dictPart = dictEntry[:length]
	} else {
		dictPart = dictEntry
	}

	// Combine all parts to form the crypto key
	return serverKey + dictPart + hostFirstChars + hostLastChar
}

// ValidateKey checks if a registration key is valid for the given serial number
func ValidateKey(serialNumber, key string) bool {
	// Base64 decode the key
	keyBytes, err := base64.StdEncoding.DecodeString(key)
	if err != nil {
		return false
	}

	// Hash the serial number to create AES key
	hasher := md5.New()
	hasher.Write([]byte(serialNumber))
	md5Key := hasher.Sum(nil)

	// Create AES cipher
	block, err := aes.NewCipher(md5Key)
	if err != nil {
		return false
	}

	// Use zero IV (16 bytes of zeros)
	iv := make([]byte, aes.BlockSize)
	mode := cipher.NewCBCDecrypter(block, iv)

	// Ensure data is a multiple of the block size
	if len(keyBytes)%aes.BlockSize != 0 {
		return false
	}

	// Decrypt data in-place
	decryptedData := make([]byte, len(keyBytes))
	mode.CryptBlocks(decryptedData, keyBytes)

	// Check if the decrypted data is the expected string
	return string(decryptedData) == "ElCloudRepSrv"
}

// GenerateKey generates a new registration key for the given serial number
func GenerateKey(serialNumber string) (string, error) {
	// The string to encrypt
	const elCloudRepSrv = "ElCloudRepSrv"

	// Hash the serial number to create AES key
	hasher := md5.New()
	hasher.Write([]byte(serialNumber))
	md5Key := hasher.Sum(nil)

	// Create AES cipher
	block, err := aes.NewCipher(md5Key)
	if err != nil {
		return "", fmt.Errorf("failed to create AES cipher: %v", err)
	}

	// Ensure data is a multiple of the block size
	paddingNeeded := aes.BlockSize - (len(elCloudRepSrv) % aes.BlockSize)
	if paddingNeeded < aes.BlockSize {
		// Add PKCS#7 padding
		paddingBytes := bytes.Repeat([]byte{byte(paddingNeeded)}, paddingNeeded)
		elCloudRepSrv += string(paddingBytes)
	}

	// Use zero IV (16 bytes of zeros)
	iv := make([]byte, aes.BlockSize)
	mode := cipher.NewCBCEncrypter(block, iv)

	// Encrypt data
	encryptedData := make([]byte, len(elCloudRepSrv))
	mode.CryptBlocks(encryptedData, []byte(elCloudRepSrv))

	// Base64 encode the encrypted data
	return base64.StdEncoding.EncodeToString(encryptedData), nil
} 